src/lib/comms/sol-connman-impl-connman.c

```diff
@@ -0,0 +1,910 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <systemd/sd-bus.h>
+#include <systemd/sd-event.h>
+
+#define SOL_LOG_DOMAIN &_sol_connman_log_domain
+#include <sol-log-internal.h>
+
+#include "sol-str-table.h"
+#include "sol-str-slice.h"
+#include "sol-util-internal.h"
+#include "sol-bus.h"
+#include "sol-connman.h"
+
+SOL_LOG_INTERNAL_DECLARE(_sol_connman_log_domain, "connman");
+
+int sol_connman_init(void);
+void sol_connman_shutdown(void);
+
+struct sol_connman_service {
+    sd_bus_slot *slot;
+    char *path;
+    char *name;
+    enum sol_connman_service_state state;
+    char *type;
+    struct sol_network_link_addr *link_addr;
+    struct sol_network_link_addr *link_addr6;
+    int32_t strength;
+    bool is_call_success;
+};
+
+struct dbus_monitor_callback {
+    void (*cb)(void *data, const struct sol_connman_service *service);
+    const void *data;
+};
+
+struct ctx {
+    struct sol_vector service_vector;
+    struct sol_vector monitor_vector;
+    struct sol_bus_client *connman;
+    sd_bus_slot *properties_changed;
+    sd_bus_slot *manager_slot;
+    sd_bus_slot *service_slot;
+    sd_bus_slot *state_slot;
+    enum sol_connman_state connman_state;
+};
+
+static struct ctx _ctx;
+
+static void
+call_monitor_callback(struct sol_connman_service *service)
+{
+    struct dbus_monitor_callback *monitor;
+    uint16_t i;
+
+    SOL_VECTOR_FOREACH_IDX (&_ctx.monitor_vector, monitor, i) {
+        if (monitor->cb)
+            monitor->cb((void *)monitor->data, service);
+    }
+}
+
+static void
+_free_connman_service(struct sol_connman_service *service)
+{
+    if (!service)
+        return;
+
+    service->slot = sd_bus_slot_unref(service->slot);
+
+    free(service->path);
+    free(service->type);
+    free(service->link_addr);
+    free(service->link_addr6);
```

Comment:
you will need a `free(service)` here and change the vector to a "ptr_vector". The reason is that the service handles will be modified once you append/remove services and users have no way to know of that, this will result in them accessing invalid pointers.